plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java

```diff
@@ -223,12 +224,23 @@ private void setupProjectSelectorDataBinding() {
         gcpProjectToProjectId, projectIdToGcpProject);
   }
 
-  private void setupTextFieldDataBinding(Control control, String modelPropertyName,
+  protected void setupTextFieldDataBinding(Text text, String modelPropertyName,
+      ValidationStatusProvider validator) {
+    ISWTObservableValue textValue = WidgetProperties.text(SWT.Modify).observe(text);
+    IObservableValue modelValue = PojoProperties.value(modelPropertyName).observe(model);
+
+    bindingContext.bindValue(textValue, modelValue);
+    if (requireValues) {
+      bindingContext.addValidationStatusProvider(validator);
+    }
```

Comment:
This doesn't seem right: if !requiresValues, then we won't validate the field contents (e.g., we could have an invalid bucket name).

Reply:
That is the point of `requireValues` that was introduced a very long time ago. The purpose is to allow the properties page (as opposed to the deploy dialog) to save invalid values (e.g., non-existing app.yaml).

BTW, this is the original code. I did not change it.

Reply:
Weird.  We should either rename it (`mustBeValidated`?) or, what I would have thought was the original intent, add support for these to be optional but valid fields?

Reply:
`mustBeValidate` sounds good. These are not about the required-optional notion.

Reply:
Where did this code come from?  I don't see it in `AppEngineDeployPreferencesPanel`.  The code I do see still does validate the fields, whereas this will only add the provided validator if `requireValues == true`.

Reply:
Oh [found it](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2411/files#diff-57c80aef5b529e610d70cdde8d3056f1L84).  Hmm, the code isn't equivalent: the previous code in `FlexDeployPreferencesPanel` was for a specific file (the app.yaml), and the validator was checking the file contents.  But this is now a utility method, and the validator may be doing other validation that is required.

Reply:
I'll tell you shortly.

BTW, I just realized this branch got a wrong merge, so I need to fix it first.

Reply:
> These are not about the required-optional notion.

That doesn't look to be the case in `AppEngineDeployPreferencesPanel`.

Reply:
This will clear your doubt: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/v1.3.1C/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java#L571

I do know the whole history of `requireValues`. It was renamed once. The value is true for the deploy dialog, and false for the property page. In the property page, we allow saving invalid values.

Reply:
Oh, I see what you're saying. Sorry, my bad. So you're thinking that validation should always be done no matter what. I was thinking skipping the validation based on the following test:

```java
  @Test
  public void testAppYamlPathValidation_noValidationIfMustBeValidatedIsFalse() {
    FlexDeployPreferencesPanel panel = createPanel(false /* mustBeValidated */);

    Text appYamlField = findAppYamlField(panel);
    appYamlField.setText("non/existing/app.yaml");
    assertNull(getAppYamlPathValidationStatus(panel));
  }
```

I was just thinking that the property page shouldn't run the validator to make it accept invalid values, but probably we should be able to make the validator just return INFO rather than ERROR. I agree your intent should be the way.